### PR TITLE
Resolve the SP-UUID from token for email branding.

### PR DESCRIPTION
**Pull request by Manchiz**
### description
Resolve the SP-UUID from token and set the SP-UUID to thread local. 

### Related issue
- https://github.com/wso2/product-is/issues/20543

### CODE REVIEW

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/OAuth2AccessTokenHandler.java [hunk 1]
```diff
@@ -187,12 +198,18 @@ protected AuthenticationResult doAuthenticate(MessageContext messageContext) {
                 }
 
                 if (serviceProvider != null) {
```
Comment (Avarjana) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/OAuth2AccessTokenHandler.java:200]: This check is done at line 179 already. I think we can remove this

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/OAuth2AccessTokenHandler.java [hunk 2]
```diff
@@ -167,11 +169,20 @@ protected AuthenticationResult doAuthenticate(MessageContext messageContext) {
...
                 try {
-                    serviceProvider =
-                            OAuth2Util.getServiceProvider(oAuth2IntrospectionResponseDTO.getClientId()).
-                                    getApplicationName();
+                    serviceProvider = OAuth2Util.getServiceProvider(oAuth2IntrospectionResponseDTO.getClientId());
+                    if (serviceProvider != null) {
+                        serviceProviderName = serviceProvider.getApplicationName();
+                        serviceProviderUUID = serviceProvider.getApplicationResourceId();
+                    } else {
+                        throw new IdentityOAuth2Exception("There is no associated Service provider for client Id "
```
Comment (Manchiz) [inline components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/OAuth2AccessTokenHandler.java:183]: add debug log

File: components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/AuthenticationValve.java [hunk 3]
```diff
@@ -273,12 +278,28 @@ private void setThreadLocalServiceProvider(AuthenticationContext authenticationC
         }
     }
 
+    private void setThreadLocalServiceProviderUUID(AuthenticationContext authenticationContext) {
```
Comment (Manchiz) [inline components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/AuthenticationValve.java:281]: use the existing method

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/OAuth2AccessTokenHandler.java [hunk 4]
```diff
@@ -167,11 +169,20 @@ protected AuthenticationResult doAuthenticate(MessageContext messageContext) {
                         OAuth2Util.buildScopeArray(oAuth2IntrospectionResponseDTO.getScope()));
                 authenticationContext.addParameter(Constants.OAUTH2_VALIDATE_SCOPE,
                         AuthConfigurationUtil.getInstance().isScopeValidationEnabled());
-                String serviceProvider = null;
+
+                ServiceProvider serviceProvider = null;
+                String serviceProviderName = null;
+                String serviceProviderUUID = null;
+
```
[the remark this answers is not in the document]
Comment (Manchiz) [reply components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/OAuth2AccessTokenHandler.java:176]: Ack

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/OAuth2AccessTokenHandler.java (hunk 2 above)
Comment (Manchiz) [reply components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/OAuth2AccessTokenHandler.java:183]: ack

File: components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/OAuth2AccessTokenHandler.java (hunk 1 above)
Comment (Manchiz) [reply components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/impl/OAuth2AccessTokenHandler.java:200]: ack

File: components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/AuthenticationValve.java (hunk 3 above)
Comment (Manchiz) [reply components/org.wso2.carbon.identity.auth.valve/src/main/java/org/wso2/carbon/identity/auth/valve/AuthenticationValve.java:281]: ack

**Review by Avarjana — APPROVED** [review]

**Pull request merged**